libraries/AP_HAL_SITL/sitl_rangefinder.cpp

```diff
@@ -44,23 +43,26 @@ void SITL_State::_update_rangefinder(float range_value)
     }
 
     float voltage = 5.0f;
-    if (fabs(_sitl->state.rollDeg) < 90 &&
-            fabs(_sitl->state.pitchDeg) < 90) {
-        // adjust for apparent altitude with roll
-        altitude /= cosf(radians(_sitl->state.rollDeg)) * cosf(radians(_sitl->state.pitchDeg));
 
-        altitude += _sitl->sonar_noise * _rand_float();
+    if (range_value == -1) {
+        if (fabs(_sitl->state.rollDeg) < 90 &&
+                fabs(_sitl->state.pitchDeg) < 90) {
+            // adjust for apparent altitude with roll
+            altitude /= cosf(radians(_sitl->state.rollDeg)) * cosf(radians(_sitl->state.pitchDeg));
+        }
+    }
 
-        // Altitude in in m, scaler in meters/volt
-        voltage = altitude / _sitl->sonar_scale;
-        voltage = constrain_float(voltage, 0, 5.0f);
+    altitude += _sitl->sonar_noise * _rand_float();
```

Comment:
I just realized that this is actually breaking previous behavior. If roll or pitch are equal or higher than 90 degrees then the voltage value should stay at 5.0f. The best piece of code I could think of is:

```
float voltage = 5.0f;
if ((fabs(_sitl->state.rollDeg) < 90 && fabs(_sitl->state.pitchDeg) < 90) || !is_equal(range_value, -1)) {
    if (is_equal(range_value, -1) {
        // adjust for apparent altitude with roll
        altitude /= cosf(radians(_sitl->state.rollDeg)) * cosf(radians(_sitl->state.pitchDeg));
    }

    altitude += _sitl->sonar_noise * _rand_float();

    // Altitude in in m, scaler in meters/volt
    voltage = altitude / _sitl->sonar_scale;
    voltage = constrain_float(voltage, 0, 5.0f);

    if (!is_zero(_sitl->sonar_glitch) && _sitl->sonar_glitch >= (_rand_float() + 1.0f) / 2.0f) {
        voltage = 5.0f;
    }
}

sonar_pin_value = 1023 * (voltage / 5.0f);
```

If you use it, then re-check that it is correct.